lib/Sema/TypeCheckDecl.cpp

```diff
@@ -1106,6 +1106,59 @@ EnumRawValuesRequest::evaluate(Evaluator &eval, EnumDecl *ED,
   return true;
 }
 
+llvm::Expected<bool>
+DidSetContainsOldValueReferenceRequest::evaluate(Evaluator &evaluator,
+                                                 AccessorDecl *decl) const {
+
+  class OldValueFinder : public ASTWalker {
+    const ASTContext &Ctx;
+    bool foundOldValueRef = false;
+
+  public:
+    OldValueFinder(const ASTContext &ctx) : Ctx(ctx) {}
+
+    virtual std::pair<bool, Expr *> walkToExprPre(Expr *E) override {
+      if (!E)
+        return {true, E};
+      if (auto DRE = dyn_cast<DeclRefExpr>(E)) {
+        auto declName = DRE->getDecl()->getBaseName();
+        if (!declName.isSpecial() && declName == Ctx.Id_oldValue) {
+          foundOldValueRef = true;
+          return {false, nullptr};
+        }
+      }
+
+      return {true, E};
+    }
+
+    bool didFindOldValueRef() { return foundOldValueRef; }
+  };
+
+  if (decl->getAccessorKind() != AccessorKind::DidSet) {
+    return false;
+  }
+
+  auto &ctx = decl->getASTContext();
+
+  // If we have an explicit didSet(argumentName), then we want to load it
+  // even if it is not used in the body of the didSet.
+  if (decl->getParameters()->size() == 1) {
+    auto param = decl->getParameters()->get(0);
+    if (!param->isImplicit()) {
+      return true;
+    }
+  }
+
+  auto walker = OldValueFinder(ctx);
+  decl->walk(walker);
+
+  auto didFindOldValue = walker.didFindOldValueRef();
+
+  // If we found a reference to the implicit 'oldValue` parameter, we want to
+  // load it. Otherwise, we can safely skip loading it.
+  return didFindOldValue;
```

Comment:
I think this can at best be a quick way of affirming that `didSet` doesn't refer to `oldValue`.  `oldValue` can be declared within the `didSet` body, perhaps even implicitly with a nested `didSet`.  We need to prove that a particular use of `oldValue` is actually referring to *this* `oldValue`, which requires us to not get tripped up by any local variables, types with accessors, and so on that might be within the `didSet` body.

Now, there's two possibilities when we're looking at the `didSet` body.  If the body has already been type-checked, we should just see `DeclRefExpr`s like this, and we can (and should) just compare the declarations instead of their names.  But if the body hasn't been type-checked — which can easily happen if e.g. the code is split across files — then we might see an `UnresolvedDeclRefExpr`, which will only have a name.  If that name is equal to the parameter name, we should abort this walk, ask the type-checker to check the body (with `swift::typeCheckAbstractFunctionBodyUntil`), and restart.  If we then continue to see UDREs, there's an error and we should just return conservatively.

Reply:
Hmm, calling `TypeChecker::typeCheckAbstractFunctionBody` inside the request leads to circularity issues.

Reply:
The accessor's `BodyKind` == `Parsed` when the request is called, so it seems like we haven't fully type checked it anyway, however any reference to `oldValue` is already a `DeclRefExpr` (even if you declare a `var oldValue` locally in the body).

Reply:
Hmm.  I was about to say that I'd talked this over with Slava and gotten convinced that it was okay to expect this request to be preconditioned on the body being type-checked, but now I'm worried about exactly how brittle that would be in terms of request dependencies.